test/instance_util.go

```diff
@@ -0,0 +1,144 @@
+// Copyright (C) 2023 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package integration
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"net"
+	"os/exec"
+	"regexp"
+	"testing"
+	"time"
+
+	"github.com/syncthing/syncthing/lib/protocol"
+	"github.com/syncthing/syncthing/lib/rand"
+)
+
+const syncthingBinary = "../bin/syncthing"
+
+// instance represents a running instance of Syncthing.
+type instance struct {
+	deviceID     protocol.DeviceID
+	syncthingDir string
+	userHomeDir  string
+	apiAddress   string
+	apiKey       string
+	tcpPort      int
+}
+
+// startInstance starts a Syncthing instance with authentication. The
+// username, password and API key are in the returned instance.
+func startInstance(t *testing.T) *instance {
+	t.Helper()
+
+	// Use temporary directories for the Syncthing and user home
+	// directories. The user home directory won't be used for anything, but
+	// it needs to exist...
+	syncthingDir := t.TempDir()
+	userHomeDir := t.TempDir()
+
+	inst := &instance{
+		syncthingDir: syncthingDir,
+		userHomeDir:  userHomeDir,
+		apiKey:       rand.String(32),
+	}
+
+	// Start Syncthing with the config and API key.
+	cmd := exec.Command(syncthingBinary, "--no-browser", "--no-default-folder", "--home", syncthingDir)
+	cmd.Env = append(basicEnv(userHomeDir), "STGUIAPIKEY="+inst.apiKey)
+	rd, wr := io.Pipe()
+	cmd.Stdout = wr
+	cmd.Stderr = wr
+	lr := newSyncthingMetadataReader(rd)
+
+	if err := cmd.Start(); err != nil {
+		t.Fatal(err)
+	}
+
+	t.Cleanup(func() {
+		cmd.Process.Kill()
+		cmd.Wait()
+	})
+
+	// Wait up to 30 seconds to get the device ID, which comes first.
+	select {
```

Comment:
minor suggestion:

I'd be inclined to bundle all the concurrency related to this operation in the one function, so you have something like

```
func readSyncthingMetadata(r io.Reader) (*syncthingMetadata, error)
```

which handles all the timeouts etc. Then you don't need the struct with all the channels, the scanner goroutine can set the fields of a syncthingMetadata struct in place as it scans them, and then when it has populated all fields it can signal to the main goroutine that it's ready to be returned.

this would make error handling more natural as well, I think